### PR TITLE
Unify the entry and insert code

### LINKED ISSUE #108: Improve .entry() performance
In some of the benchmarks, if I replace `.insert()` with code based on entry, the benchmarks run slower. If we could get these to performance parity, then we can delete the `.insert()` code, since it is duplicated by the functionality in `.entry()`.

To test this, you can update insert to be like this (note `insert_full` uses `entry` right now):

```rust
    pub fn insert(&mut self, key: K, value: V) -> Option<V> {
        self.insert_full(key, value).1
    }
```

The following is the benchmark difference I see when using `entry` in `insert`.

The hashmap benchmarks can be ignored - ordermap is the interesting part. (The hashmap benchmarks are controls.)

Change from before to after. 
```
 name                                   63 ns/iter  62 ns/iter  diff ns/iter  diff % 
 insert_hashmap_100_000                 2,488,863   2,504,113         15,250   0.61% 
 insert_hashmap_10_000                  214,543     221,693            7,150   3.33% 
 insert_hashmap_150                     3,210       3,339                129   4.02% 
 insert_hashmap_int_bigvalue_10_000     262,701     264,518            1,817   0.69% 
 insert_hashmap_str_10_000              236,377     247,805           11,428   4.83% 
 insert_hashmap_string_10_000           1,117,104   1,148,170         31,066   2.78% 
 insert_hashmap_string_oneshot_10_000   1,078,875   1,085,733          6,858   0.64% 
 insert_orderedmap_100_000              2,727,119   3,874,446      1,147,327  42.07% 
 insert_orderedmap_10_000               267,361     363,741           96,380  36.05% 
 insert_orderedmap_150                  3,956       5,664              1,708  43.17% 
 insert_orderedmap_int_bigvalue_10_000  359,582     413,168           53,586  14.90% 
 insert_orderedmap_str_10_000           305,463     312,993            7,530   2.47% 
 insert_orderedmap_string_10_000        1,066,189   1,091,526         25,337   2.38%
```

### PR DESCRIPTION
**Pull request by mwillsey**
This will close #108 

This doesn't so much improve the entry code rather than abstract away `entry` and `insert` into a `ProbeAction`. This forces monomorphization to give us the insert fast path that doesn't require the additional indirection that entry does.

It also implements `insert_full` on that fast path (`insert` is now just a call to `insert_full`), so that should be much faster as well, since it was formerly implemented in terms of `entry`.

Benchmarks
```
 name                                   control ns/iter  variable ns/iter  diff ns/iter  diff %  speedup 
 insert_orderedmap_100_000              2,688,143        2,616,739              -71,404  -2.66%   x 1.03 
 insert_orderedmap_10_000               275,198          272,283                 -2,915  -1.06%   x 1.01 
 insert_orderedmap_150                  4,116            3,862                     -254  -6.17%   x 1.07 
 insert_orderedmap_int_bigvalue_10_000  409,681          395,290                -14,391  -3.51%   x 1.04 
 insert_orderedmap_str_10_000           298,722          294,373                 -4,349  -1.46%   x 1.01 
 insert_orderedmap_string_10_000        2,027,784        1,948,038              -79,746  -3.93%   x 1.04 
```

### CODE REVIEW

File: src/map.rs [hunk 1]
```diff
@@ -1785,6 +1694,59 @@ impl<K, V> OrderMapCore<K, V> {
...
+        *pos = Pos::with_hash::<Sz>(index, entry.hash);
+        entry.map.entries.push(Bucket {
+            hash: entry.hash,
+            key: entry.key,
+            value: self.0,
+        });
+        (index, None)
+    }
+    fn steal(self, entry: VacantEntry<'a, K, V>) -> Self::Output {
+        let index = entry.map.entries.len();
```
Comment (bluss) [inline src/map.rs:1726]: this could be just `entry.insert_impl::<Sz>(self.0)` right? Something like that to reduce duplication

**Comment by bluss** [comment]
That's awesome

File: src/map.rs [hunk 2]
```diff
@@ -1785,6 +1694,59 @@ impl<K, V> OrderMapCore<K, V> {
     }
 }
 
+trait ProbeAction<'a, Sz: Size, K, V>: Sized {
+    type Output;
+    fn hit(self, entry: OccupiedEntry<'a, K, V>) -> Self::Output;
```
Comment (bluss) [inline src/map.rs:1699]: If possible, document each method with a one-line summary of what it represents

File: src/map.rs [hunk 3]
```diff
@@ -1785,6 +1694,59 @@ impl<K, V> OrderMapCore<K, V> {
...
+
+struct InsertValue<V>(V);
+
+impl<'a, Sz: Size, K, V> ProbeAction<'a, Sz, K, V> for InsertValue<V> {
+    type Output = (usize, Option<V>);
+    fn hit(self, entry: OccupiedEntry<'a, K, V>) -> Self::Output {
+        let old = replace(&mut entry.map.entries[entry.index].value, self.0);
+        (entry.index, Some(old))
+    }
+    fn empty(self, entry: VacantEntry<'a, K, V>) -> Self::Output {
```
Comment (bluss) [inline src/map.rs:1714]: A style nitpick is that all the methods need an empty line separating them from the item before them

File: src/map.rs [hunk 4]
```diff
@@ -840,13 +834,13 @@ where
     K: Hash + Eq,
     S: BuildHasher,
 {
-    // FIXME: reduce duplication (compare with insert)
-    fn entry_phase_1<Sz>(&mut self, key: K) -> Entry<K, V>
+    fn probe_action<'a, Sz, A>(&'a mut self, key: K, action: A) -> A::Output
```
Comment (bluss) [inline src/map.rs:837]: Nice refactoring. I think this method needs a clearer name. We use "probe" in every loop when reading/writing to the map, so it is not specific enough. Even "insert phase 1" is an ok name - and `insert_phase_2` still exists, so it needs to be renamed in tandem, if we entirely remove anything named "phase 1".

File: src/map.rs [hunk 5]
```diff
@@ -1785,6 +1694,59 @@ impl<K, V> OrderMapCore<K, V> {
     }
 }
 
+trait ProbeAction<'a, Sz: Size, K, V>: Sized {
+    type Output;
+    fn hit(self, entry: OccupiedEntry<'a, K, V>) -> Self::Output;
+    fn empty(self, entry: VacantEntry<'a, K, V>) -> Self::Output;
+    fn steal(self, entry: VacantEntry<'a, K, V>) -> Self::Output {
+        self.empty(entry)
+    }
```
Comment (bluss) [inline src/map.rs:1703]: For code readability we can prefer to not have this default impl for `steal`.

File: src/map.rs (hunk 1 above)
Comment (mwillsey) [reply src/map.rs:1726]: Ah good idea. Do you mean `empty` could use that? I was afraid that entering the probe loop another time would be costly (technically, you don't have to in the empty case), but it seems to be fine.

Comment (mwillsey) [reply src/map.rs:1730]: Nvm, turns out doing it in `empty` isn't great, it basically remove the fast path that the old insert code had, making steal and empty the same.

Comment (bluss) [reply src/map.rs:1730]: Are we then maybe identifying the main thing that slows down entry compared with the old insert? Maybe if we updated entry's VacantEntry to skip that loop, for the non-steal case, that insert and entry would match each other in performance?

**Comment by bluss** [comment]
Could you update the PR description to use one of the formulations that means that #108 is closed when this is merged? :) https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

File: src/map.rs [hunk 6]
```diff
@@ -1785,6 +1694,60 @@ impl<K, V> OrderMapCore<K, V> {
     }
 }
 
+trait ProbeAction<'a, Sz: Size, K, V>: Sized {
+    type Output;
+    // handle an occupied spot in the map
+    fn hit(self, entry: OccupiedEntry<'a, K, V>) -> Self::Output;
+    // handle an empty spot in the map
+    fn empty(self, entry: VacantEntry<'a, K, V>) -> Self::Output;
+    // robin hood: handle a that you should steal because it's better for you
```
Comment (cuviper) [inline src/map.rs:1703]:
```suggestion
    // robin hood: handle a spot that you should steal because it's better for you
```

**Comment by bluss** [comment]
Thanks. I'll merge this when I have time to make some post-merge edits

**Comment by cuviper** [comment]
@bluss can I help with whatever is blocking this?

**Comment by bluss** [comment]
I'll add it as comments. I thought I might as well merge and edit myself, sometimes I prefer that to the back-and-forth, just don't want reviews to be too perfectionist

File: src/map.rs [hunk 7]
```diff
@@ -1785,6 +1694,60 @@ impl<K, V> OrderMapCore<K, V> {
...
+    fn empty(self, entry: VacantEntry<'a, K, V>) -> Self::Output {
+        let pos = &mut entry.map.indices[entry.probe];
+        let index = entry.map.entries.len();
+        *pos = Pos::with_hash::<Sz>(index, entry.hash);
+        entry.map.entries.push(Bucket {
+            hash: entry.hash,
+            key: entry.key,
+            value: self.0,
+        });
+        (index, None)
```
Comment (bluss) [inline src/map.rs:1726]: This method body seems "loose" like this - could it not just call a method on the entry? I general, it's hard to follow the insert logic when it's spread out.

(This comment is rather perfectionist, and not something I wanted to make a requirement to fix before merging).

File: src/map.rs [hunk 8]
```diff
@@ -1785,6 +1694,60 @@ impl<K, V> OrderMapCore<K, V> {
...
+}
+
+struct MakeEntry;
+
+impl<'a, Sz: Size, K: 'a, V: 'a> ProbeAction<'a, Sz, K, V> for MakeEntry {
+    type Output = Entry<'a, K, V>;
+    fn hit(self, entry: OccupiedEntry<'a, K, V>) -> Self::Output {
+        Entry::Occupied(entry)
+    }
+    fn empty(self, entry: VacantEntry<'a, K, V>) -> Self::Output {
```
Comment (bluss) [inline src/map.rs:1743]: nitpick missing empty line between the methods

**Comment by cuviper** [comment]
Yeah, I get that, and sometimes I'll take advantage of maintainer pushes to PRs for that same reason.

I'm pressing because I'm building up changes toward [using indexmap more in rustc](https://github.com/rust-lang/rust/issues/60608). :slightly_smiling_face:

File: src/map.rs (hunk 7 above)
Comment (cuviper) [reply src/map.rs:1726]: I looked at this -- you'd basically want `VacantEntry::insert_impl` without the probe loop (since we know its empty). But this needs to consume the entry's key, which makes it harder to reuse that code, so I think we'd end up writing this code separately either way. i.e. It doesn't really matter whether we do it here or there.

(And frankly, a lot of this will disappear if I land the hashbrown changes...)

**Comment by cuviper** [comment]
I fixed the spacing nit, and I think we can wait on the other point.

Thanks @mwillsey!
